internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -183,9 +183,28 @@ object ClassToAPI {
     val fields = mergeMap(c, c.getDeclaredFields, c.getFields, fieldToDef(c, cf, enclPkg))
     val constructors =
       mergeMap(c, c.getDeclaredConstructors, c.getConstructors, constructorToDef(enclPkg))
+    val cl = c.getClassLoader
+    val name = c.getCanonicalName
+    val innerClasses = new mutable.ArrayBuffer[Class[_]]()
+    val publicClasses = new mutable.ArrayBuffer[Class[_]]()
+    val it = cf.innerClasses.toIterator
+    while (it.hasNext) {
+      val info = it.next
+      try {
+        if (info.outerClassName == name) {
+          val innerClass = cl.loadClass(info.innerClassName)
+          innerClasses += innerClass
+          if (info.isPublic) {
+            publicClasses += innerClass
+          }
+        }
+      } catch {
+        case _: ClassNotFoundException => // do nothing
```

Comment:
@jvican 

> 2. Do we have any idea what we're after here? Why is this exception being thrown? Are we now not throwing this exception because we're potentially doing less work, which could affect the incremental compiler for java sources?

The case that we have repro project (https://github.com/duboisf/sbt_issue_117_example) is a simple case of `java.lang.NoClassDefFoundError: org/jboss/virtual/VirtualFileVisitor` while loading some class. Why is this happening? Look no further than spring-core 3.0.0.RELEASE's POM file:

http://central.maven.org/maven2/org/springframework/spring-core/3.0.0.RELEASE/spring-core-3.0.0.RELEASE.pom

```xml
<dependency>
  <groupId>org.jboss.vfs</groupId>
  <artifactId>com.springsource.org.jboss.virtual</artifactId>
  <version>2.1.0.GA</version>
  <scope>compile</scope>
  <optional>true</optional>
</dependency>
```

Maven allows optional dependency, and Spring optionally depends on JBoss VFS. We can find that some of the inner classes in Spring Core exposes `org.jboss.virtual.VirtualFileVisitor` as parent class. Based on the release date, it's likely this code here: https://github.com/spring-projects/spring-framework/blob/784068346df62225661d9ce1bbd1a32a1fddde66/org.springframework.core/src/main/java/org/springframework/core/io/support/PathMatchingResourcePatternResolver.java#L662:

```java
import org.jboss.virtual.VirtualFileVisitor;
...
	private static class PatternVirtualFileVisitor implements VirtualFileVisitor {
```

Zinc is written to keep tracking parent classes, and doing to requires loading the class. I am doing the same, but I am just ignoring the `ClassNotFoundException`.

Reply:
> 3. Have we tried this in different JVM versions?

No, but optional JAR missing from classpath is a problem that's independent from JVM versions. Since it's reported in 2011 and 2014, likely it covers Java 6, 7, and 8.